### PR TITLE
VS Code extension: Provide smart contract snippets

### LINKED ISSUE #37: Provide code snippets for common smart contract patterns (VS Code extension)
The extension should include useful snippets.

Depending on the user experience and expressiveness of these snippets, details will have to be decided as these are developed.

Ex. Defining init and receive functions for smart contracts involve some amount of boilerplate, and the extension should include snippets to reduce some of the work related to this.

Snippets to implement:
- [x] `#[init(..)]`
- [x] `#[receive(..)]`

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

Closes #37 

## Changes

Include a number of snippets for smart contract development

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [ ] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: vscode-smart-contracts/snippets.json [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  "Smart contract state struct": {
+    "prefix": ["scstate"],
+    "body": [
+      "#[derive(Serial, DeserialWithState, StateClone)]",
+      "#[concordium(state_parameter = \"S\")]",
+      "struct ${1:MyContractState}<S> {",
+      "\t$0",
+      "}"
+    ],
+    "description": "State struct for a Concordium smart contract"
```
Comment (Bargsteen) [inline vscode-smart-contracts/snippets.json:50]: Perhaps we should mention that the generic parameter Is only needed if you state* types in there.

File: vscode-smart-contracts/snippets.json [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+      "impl From<LogError> for $1 {",
+      "\tfn from(log_error: LogError) -> Self {",
+      "\t\tmatch log_error {",
+      "\t\t\tLogError::Full => Self::LogFull,",
+      "\t\t\tLogError::Malformed => Self::LogMalformed,",
+      "\t\t}",
+      "\t}",
+      "}"
+    ],
+    "description": "Error type for a Concordium smart contract"
```
Comment (Bargsteen) [inline vscode-smart-contracts/snippets.json:73]: Nice!

**Review by Bargsteen — APPROVED** [review]
Looks really good and useful! I've only read it on GitHub. I'll test it out in VSCode on Monday.

File: vscode-smart-contracts/snippets.json (hunk 1 above)
Comment (limemloh) [reply vscode-smart-contracts/snippets.json:50]: This description is not really shown to the user (only if they run a command for listing every snippet), so we can't really educate them here. 
Maybe the snippet should be without the generic?

Comment (Bargsteen) [reply vscode-smart-contracts/snippets.json:50]: Ah okay.
Hmm, I am not sure about that.
On the one hand, we have this epic about removing the generics, which suggests you remove it from the snippet.
But on the other, the generic part is still being used and is harder to remember and type, so having a snippet for it is currently useful.

File: vscode-smart-contracts/snippets.json [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+  },
+  "Smart contract test receive function": {
+    "prefix": ["sctestrec"],
+    "body": [
+      "#[concordium_test]",
+      "fn test_$1() {",
+      "\tlet ctx = TestReceiveContext::empty();",
+      "\tlet state = $2;",
+      "\tlet host = TestHost::new(state, TestStateBuilder::new())",
```
Comment (Bargsteen) [inline vscode-smart-contracts/snippets.json:82]:
```suggestion
      "\tlet host = TestHost::new(state, TestStateBuilder::new());",
```

**Comment by Bargsteen** [comment]
I just tried the snippets and they work very well!
Have you looked into which prefix other blockchains use in their vscode snippets?
At least we should make sure not to clash with any of the major ones.
We can consider using `ccd*` if `sc*` is already taken.

**Review by abizjak — APPROVED** [review]

**Review by DOBEN — APPROVED** [review]

File: vscode-smart-contracts/snippets.json [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  "Smart contract init-function": {
+    "prefix": ["scinit"],
+    "body": [
+      "#[init(",
+      "\tcontract = \"${1:my-contract}\",",
+      "\tparameter = \"${2:MyParams}\",",
+      "\terror = \"${3:MyContractError}\",",
+      "\tevent = \"${4:MyContractEvent}\"",
+      ")]",
+      "fn $1_init<S: HasStateApi>(",
```
[the remark this answers is not in the document]
Comment (limemloh) [reply vscode-smart-contracts/snippets.json:31]: It is possible yes, they even have an example where they replace - with _

File: vscode-smart-contracts/snippets.json (hunk 1 above)
Comment (limemloh) [reply vscode-smart-contracts/snippets.json:50]: I'll leave it as is for now.

**Comment by limemloh** [comment]
> I just tried the snippets and they work very well! Have you looked into which prefix other blockchains use in their vscode snippets? At least we should make sure not to clash with any of the major ones. We can consider using `ccd*` if `sc*` is already taken.

I did not find any other chains with clashing prefix, but I have changed it to `ccd` as you suggested. It seems to be the safer choice.

**Pull request merged**